File: ios/Sources/Log4Agent/Log4Agent.swift
import Foundation

public final class Log4Agent {
    public static let shared = Log4Agent()

    private var config = Log4AgentConfig()
    private var session: URLSession = .shared
    private let encoder = JSONEncoder()
    private var sessionId = UUID().uuidString
    private var sessionStarted = false
    private var initialized = false

    public init() {}

    public func configure(_ config: Log4AgentConfig = Log4AgentConfig(), session: URLSession = .shared, force: Bool = false) {
        if initialized && !force { return }
        initialized = true
        self.config = config
        self.session = session
        self.sessionId = UUID().uuidString
        self.sessionStarted = false
        startSession()
    }

    public func debug(_ category: String, _ message: String, attributes: [String: String] = [:]) {
        log(category, message, level: .debug, attributes: attributes)
    }

    public func info(_ category: String, _ message: String, attributes: [String: String] = [:]) {
        log(category, message, level: .info, attributes: attributes)
    }

    public func warn(_ category: String, _ message: String, attributes: [String: String] = [:]) {
        log(category, message, level: .warn, attributes: attributes)
    }

    public func error(_ category: String, _ message: String, attributes: [String: String] = [:]) {
        log(category, message, level: .error, attributes: attributes)
    }

    public func log(
        _ category: String,
        _ message: String,
        level: Log4AgentLevel = .info,
        attributes: [String: String] = [:]
    ) {
        let current = config
        guard current.enabled else { return }
        startSession()
        let safeMessage = current.redactionEnabled
            ? Log4AgentRedactor.redactURL(message, keys: current.redactionKeys)
            : message
        let safeAttributes = current.redactionEnabled
            ? Log4AgentRedactor.redactAttributes(attributes, keys: current.redactionKeys)
            : attributes
        let payload = Log4AgentPayload(
            timestamp: ISO8601DateFormatter().string(from: Date()),
            app: current.app,
            deviceId: current.deviceId,
            sessionId: sessionId,
            platform: "ios",
            level: level.rawValue,
            category: category,
            message: safeMessage,
            attributes: safeAttributes
        )
        guard let data = try? encoder.encode(payload) else { return }
        var request = URLRequest(url: current.endpoint)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "content-type")
        request.httpBody = data
        session.dataTask(with: request).resume()
    }

    private func startSession() {
        let current = config
        guard current.enabled, !sessionStarted else { return }
        sessionStarted = true
        let payload = Log4AgentSessionPayload(
            timestamp: ISO8601DateFormatter().string(from: Date()),
            app: current.app,
            deviceId: current.deviceId,
            sessionId: sessionId,
            platform: "ios"
        )
        guard let data = try? encoder.encode(payload) else { return }
        var request = URLRequest(url: sessionEndpoint(current.endpoint))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "content-type")
        request.httpBody = data
        session.dataTask(with: request).resume()
    }

    private func sessionEndpoint(_ logEndpoint: URL) -> URL {
        let absolute = logEndpoint.absoluteString
        if absolute.hasSuffix("/logs") {
            return URL(string: String(absolute.dropLast("/logs".count)) + "/sessions")!
        }
        return URL(string: absolute.trimmingCharacters(in: CharacterSet(charactersIn: "/")) + "/sessions")!
    }
}

private struct Log4AgentPayload: Encodable {
    let timestamp: String
    let app: String
    let deviceId: String
    let sessionId: String
    let platform: String
    let level: String
    let category: String
    let message: String
    let attributes: [String: String]
}

private struct Log4AgentSessionPayload: Encodable {
    let timestamp: String
    let app: String
    let deviceId: String
    let sessionId: String
    let platform: String
}

File: android/log4agent/src/main/java/dev/log4agent/Log4Agent.kt
package dev.log4agent

import okhttp3.Call
import okhttp3.Callback
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import org.json.JSONObject
import java.io.IOException
import java.time.Instant
import java.util.UUID

object Log4Agent {
    private val jsonMediaType = "application/json; charset=utf-8".toMediaType()
    private var config = Log4AgentConfig()
    private var client = OkHttpClient()
    private var sessionId = UUID.randomUUID().toString()
    private var sessionStarted = false

    fun configure(config: Log4AgentConfig = Log4AgentConfig(), okHttpClient: OkHttpClient? = null) {
        this.config = config
        sessionId = UUID.randomUUID().toString()
        sessionStarted = false
        if (okHttpClient != null) client = okHttpClient
        startSession()
    }

    fun debug(category: String, message: String, attributes: Map<String, String> = emptyMap()) =
        log(category, message, Log4AgentLevel.Debug, attributes)

    fun info(category: String, message: String, attributes: Map<String, String> = emptyMap()) =
        log(category, message, Log4AgentLevel.Info, attributes)

    fun warn(category: String, message: String, attributes: Map<String, String> = emptyMap()) =
        log(category, message, Log4AgentLevel.Warn, attributes)

    fun error(category: String, message: String, attributes: Map<String, String> = emptyMap()) =
        log(category, message, Log4AgentLevel.Error, attributes)

    fun log(
        category: String,
        message: String,
        level: Log4AgentLevel = Log4AgentLevel.Info,
        attributes: Map<String, String> = emptyMap(),
    ) {
        val current = config
        if (!current.enabled || current.endpoint.isBlank()) return
        startSession()
        val safeMessage = if (current.redactionEnabled) Log4AgentRedactor.redactUrl(message, current.redactionKeys) else message
        val safeAttributes = if (current.redactionEnabled) Log4AgentRedactor.redactAttributes(attributes, current.redactionKeys) else attributes
        val body = buildJson(category, safeMessage, level, safeAttributes)
        val requestBuilder = Request.Builder()
            .url(current.endpoint)
            .post(body.toRequestBody(jsonMediaType))
            .header("content-type", "application/json")
        client.newCall(requestBuilder.build()).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) = Unit
            override fun onResponse(call: Call, response: Response) {
                response.close()
            }
        })
    }

    private fun startSession() {
        val current = config
        if (sessionStarted || !current.enabled || current.endpoint.isBlank()) return
        sessionStarted = true
        val body = JSONObject()
            .put("timestamp", Instant.now().toString())
            .put("app", current.app)
            .put("deviceId", current.deviceId)
            .put("sessionId", sessionId)
            .put("platform", "android")
            .toString()
        val request = Request.Builder()
            .url(sessionEndpoint(current.endpoint))
            .post(body.toRequestBody(jsonMediaType))
            .header("content-type", "application/json")
            .build()
        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) = Unit
            override fun onResponse(call: Call, response: Response) {
                response.close()
            }
        })
    }

    private fun sessionEndpoint(logEndpoint: String): String =
        if (logEndpoint.endsWith("/logs")) logEndpoint.removeSuffix("/logs") + "/sessions"
        else logEndpoint.trimEnd('/') + "/sessions"

    private fun buildJson(
        category: String,
        message: String,
        level: Log4AgentLevel,
        attributes: Map<String, String>,
    ): String {
        return JSONObject()
            .put("timestamp", Instant.now().toString())
            .put("app", config.app)
            .put("deviceId", config.deviceId)
            .put("sessionId", sessionId)
            .put("platform", "android")
            .put("level", level.name.lowercase())
            .put("category", category)
            .put("message", message)
            .put("attributes", JSONObject(attributes))
            .toString()
    }
}
